PL2/PL2-shared/src/main/java/nl/tudelft/pl2016gr2/model/BaseSequence.java

```diff
@@ -0,0 +1,170 @@
+package nl.tudelft.pl2016gr2.model;
+
+import nl.tudelft.pl2016gr2.thirdparty.testing.utility.TestId;
+
+/**
+ * This is a container class for the bases of a genome. It efficiently encodes the ATCG and N
+ * (unknown) bases as bits in an array of integers, so fewer memory is needed than when using a
+ * simple String.
+ *
+ * @author Faris
+ */
+public class BaseSequence {
+
+  private static final int BITS_PER_BASE = 3;
+  private static final int BITS_PER_INT = Integer.BYTES * 8;
+
+  private static final int END_OF_BASES = 0b000;
+  private static final int BASE_A = 0b001;
+  private static final int BASE_T = 0b010;
+  private static final int BASE_C = 0b011;
+  private static final int BASE_G = 0b100;
+  private static final int BASE_UNKNOWN = 0b101;
+
+  @TestId(id = "bases")
+  private final int[] bases;
+
+  /**
+   * Create a base sequence.
+   *
+   * @param bases the sequence of bases.
+   */
+  public BaseSequence(String bases) {
+    this(bases.toCharArray(), 0, bases.length());
+  }
+
+  /**
+   * Create a base sequence.
+   *
+   * @param bases      an array containing the sequence of bases.
+   * @param startIndex the index in the array where the bases start.
+   * @param endIndex   the index in the array where the bases end.
+   */
+  public BaseSequence(char[] bases, int startIndex, int endIndex) {
+    int amountOfBases = endIndex - startIndex;
+    int bitsNeeded = amountOfBases * BITS_PER_BASE;
+    int intsNeeded = (int) Math.ceil(bitsNeeded / (double) BITS_PER_INT);
+    this.bases = new int[intsNeeded];
+    setBases(bases, startIndex, endIndex);
+  }
+
+  /**
+   * Get the base sequence as a string.
+   *
+   * @return the base sequence as a string.
+   */
+  public String getBaseSequence() {
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; (i * BITS_PER_BASE) / BITS_PER_INT < bases.length; i++) {
+      int encodedBase = getBase((i * BITS_PER_BASE) / BITS_PER_INT,
+          (i * BITS_PER_BASE) % BITS_PER_INT);
+      if (encodedBase == END_OF_BASES) {
+        break;
+      } else {
+        sb.append(decode(encodedBase));
+      }
+    }
+    return sb.toString();
+  }
+
+  /**
+   * Get a single encoded base.
+   *
+   * @param arrayIndex the index in the array of the base.
+   * @param bitIndex   the index of the starting bit of the base in the array.
+   * @return the read encoded base.
+   */
+  private int getBase(int arrayIndex, int bitIndex) {
+    if (bitIndex <= BITS_PER_INT - BITS_PER_BASE) {
+      return (bases[arrayIndex] >> bitIndex) & 0b111;
+    } else if (arrayIndex + 1 < bases.length) {
+      int bits = (bases[arrayIndex] >>> bitIndex) & 0b111;
```

Comment:
I had a guess that this would go wrong for `>>` because of shifting in `1` bits. just to be sure, could you test it to see if it works with the signed shift? If so, it would be more consistent. Otherwise, obviously the unsigned shift is required and that be it.